### PR TITLE
Remove duplicate features using sanity checker feature to feature correlations

### PR DESCRIPTION
**Pull request by leahmcguire**
**Related issues**
Refer to issue(s) addressed in this pull request from [Issues](https://github.com/salesforce/TransmogrifAI/issues) page.

**Describe the proposed solution**
When features are exact duplicates of each other we would like to remove one before modeling

**Describe alternatives you've considered**

**Additional context**
Add any other context about the changes here.

### CODE REVIEW

File: core/src/main/scala/com/salesforce/op/stages/impl/preparators/SanityCheckerMetadata.scala [hunk 1]
```diff
@@ -319,29 +323,45 @@ case class Correlations
...
   private def correlationsFromMetadata(meta: Metadata): Correlations = {
     val wrapped = meta.wrapped
-    Correlations(
-      featuresIn = wrapped.getArray[String](SanityCheckerNames.FeaturesIn).toSeq,
-      values = wrapped.getArray[Double](SanityCheckerNames.Values).toSeq,
-      nanCorrs = wrapped.getArray[String](SanityCheckerNames.CorrelationsWLabelIsNaN).toSeq,
-      corrType = CorrelationType.withNameInsensitive(wrapped.get[String](SanityCheckerNames.CorrelationType))
-    )
+    val features = wrapped.getArray[String](SanityCheckerNames.FeaturesIn).toSeq
+    if (wrapped.underlyingMap.keySet.contains("values")) { // old sanity checker meta
```
Comment (nicodv) [inline core/src/main/scala/com/salesforce/op/stages/impl/preparators/SanityCheckerMetadata.scala:345]: nice backwards compatibility hack! maybe use "correlationsWithLabel" instead of "values", as its speaks more to the type of change that occurred?

File: core/src/main/scala/com/salesforce/op/stages/impl/preparators/SanityCheckerMetadata.scala [hunk 2]
```diff
@@ -274,23 +273,24 @@ case class CategoricalStats
...
 (
   featuresIn: Seq[String],
-  values: Seq[Double],
-  nanCorrs: Seq[String],
+  valuesWithLabel: Seq[Double],
+  valuesWithFeatures: Seq[Seq[Double]],
   corrType: CorrelationType
 ) extends MetadataLike {
-  require(featuresIn.length == values.length, "Feature names and correlation values arrays must have the same length")
+  require(featuresIn.length == valuesWithLabel.length,
```
[the remark this answers is not in the document]
Comment (leahmcguire) [reply core/src/main/scala/com/salesforce/op/stages/impl/preparators/SanityCheckerMetadata.scala:287]: correlation matrix may be empty so not sure it makes sense

File: core/src/main/scala/com/salesforce/op/stages/impl/preparators/SanityChecker.scala [hunk 3]
```diff
@@ -94,15 +94,24 @@ trait SanityCheckerParams extends DerivedFeatureFilterParams {
   final val maxCorrelation = new DoubleParam(
     parent = this, name = "maxCorrelation",
     doc = "Maximum correlation (absolute value) allowed between a feature in the feature vector and the label",
-    isValid = ParamValidators.inRange(lowerBound = 0.0, upperBound = 1.0, lowerInclusive = true, upperInclusive = true)
+    isValid = ParamValidators.inRange(lowerBound = -0.1, upperBound = 1.1, lowerInclusive = true, upperInclusive = true)
```
[the remark this answers is not in the document]
Comment (leahmcguire) [reply core/src/main/scala/com/salesforce/op/stages/impl/preparators/SanityChecker.scala:97]: Floating point - I found in some tests that if you just have a check that the corr be > 1.0 doubles mean that some perfectly correlated features are excluded because they have a corr of 1.00000000004 an I wanted to be able to turn the check off effectively without adding another param

File: core/src/main/scala/com/salesforce/op/stages/impl/preparators/DerivedFeatureFilterUtils.scala [hunk 4]
```diff
@@ -174,6 +176,10 @@ object DerivedFeatureFilterUtils {
             case ind => Option(corrsWithLabel(ind))
           },
           cramersV = cramersVMap.get(name),
+          featureCorrs = corrIndices.indexOf(i) match {
```
[the remark this answers is not in the document]
Comment (leahmcguire) [reply core/src/main/scala/com/salesforce/op/stages/impl/preparators/DerivedFeatureFilterUtils.scala:179]: So this line is part of creating the exclusion criteria by row it cannot go outside the iteration over i

**Comment by nicodv** [comment]
LGTM

File: core/src/main/scala/com/salesforce/op/stages/impl/preparators/SanityChecker.scala (hunk 3 above)
Comment (Jauntbox) [inline core/src/main/scala/com/salesforce/op/stages/impl/preparators/SanityChecker.scala:97]: Why did these ranges need to be changed. We shouldn't get correlations outside of [0, 1], right?

File: core/src/main/scala/com/salesforce/op/stages/impl/preparators/SanityChecker.scala [hunk 5]
```diff
@@ -94,15 +94,24 @@ trait SanityCheckerParams extends DerivedFeatureFilterParams {
...
+  )
+  def setMaxFeatureCorr(value: Double): this.type = set(maxFeatureCorr, value)
+  def getMaxFeatureCorr: Double = $(maxFeatureCorr)
+
   final val minCorrelation = new DoubleParam(
     parent = this, name = "minCorrelation",
     doc = "Minimum correlation (absolute value) allowed between a feature in the feature vector and the label",
-    isValid = ParamValidators.inRange(lowerBound = 0.0, upperBound = 1.0, lowerInclusive = true, upperInclusive = true)
+    isValid = ParamValidators.inRange(lowerBound = -0.1, upperBound = 1.1, lowerInclusive = true, upperInclusive = true)
```
Comment (Jauntbox) [inline core/src/main/scala/com/salesforce/op/stages/impl/preparators/SanityChecker.scala:114]: Same here - why are these changed?

File: core/src/main/scala/com/salesforce/op/stages/impl/preparators/SanityCheckerMetadata.scala [hunk 6]
```diff
@@ -92,9 +92,8 @@ case class SanityCheckerSummary
     sample: Double
   ) {
     this(
-      correlationsWLabel = new Correlations(
-        stats.filter(s => s.corrLabel.isDefined && !s.corrLabel.get.isNaN).map(s => s.name -> s.corrLabel.get),
-        stats.filter(s => s.corrLabel.isDefined && s.corrLabel.get.isNaN).map(_.name),
+      correlations = new Correlations(
+        stats.filter(s => s.corrLabel.isDefined).map(s => (s.name, s.corrLabel.get, s.featureCorrs)),
```
Comment (Jauntbox) [inline core/src/main/scala/com/salesforce/op/stages/impl/preparators/SanityCheckerMetadata.scala:96]: Do we not have to check for NaNs anymore? What if the feature is constant and we turn off the min variance threshold?

File: core/src/main/scala/com/salesforce/op/stages/impl/preparators/SanityCheckerMetadata.scala [hunk 7]
```diff
@@ -274,23 +271,24 @@ case class CategoricalStats
  * Correlations between features and the label from [[SanityChecker]]
  *
  * @param featuresIn names of features
- * @param values     correlation of feature with label
- * @param nanCorrs   nan correlation features
+ * @param valuesWithLabel     correlation of feature with label
+ * @param valuesWithFeatures   correlations between features features
```
Comment (Jauntbox) [inline core/src/main/scala/com/salesforce/op/stages/impl/preparators/SanityCheckerMetadata.scala:275]: "correlations between features ~~features~~"

File: core/src/main/scala/com/salesforce/op/stages/impl/preparators/SanityChecker.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (nicodv) [reply core/src/main/scala/com/salesforce/op/stages/impl/preparators/SanityChecker.scala:97]: @Jauntbox ^^^

File: core/src/test/scala/com/salesforce/op/stages/impl/preparators/BadFeatureZooTest.scala [hunk 8]
```diff
@@ -88,6 +88,7 @@ class BadFeatureZooTest extends FlatSpec with TestSparkContext with Logging {
     val genFeatureVector = Seq(rawCity, rawCountry, rawPickList, rawCurrency).transmogrify()
     val checkedFeatures = new SanityChecker()
       .setCheckSample(1.0)
+      .setMaxFeatureCorr(1.1)
```
Comment (Jauntbox) [inline core/src/test/scala/com/salesforce/op/stages/impl/preparators/BadFeatureZooTest.scala:91]: Oh, I see - you're using values > 1.0 to turn it off. I think it may be clearer if we keep the bounds in [0, 1.0], but make it a < check instead of a <= check (so that setting it to 1.0 would turn it off).

Comment (leahmcguire) [reply core/src/test/scala/com/salesforce/op/stages/impl/preparators/BadFeatureZooTest.scala:91]: So I did that! Turns out that with doubles you can get a correlation > 1.0 with perfectly correlated things :-P

Eg in a couple of test 1 had it set to 1.0 with corr > and it failed because the correlation was 1.00000000004

File: core/src/main/scala/com/salesforce/op/stages/impl/preparators/SanityCheckerMetadata.scala (hunk 6 above)
Comment (leahmcguire) [reply core/src/main/scala/com/salesforce/op/stages/impl/preparators/SanityCheckerMetadata.scala:96]: So I fixed it - in that when it goes to metadata it turns it into strings and then when it gets out back into the class it is turned into doubles. Now nans work! The json serializer would not have helped because it was a problem with the spark metadata.

File: core/src/main/scala/com/salesforce/op/stages/impl/preparators/SanityChecker.scala [hunk 9]
```diff
@@ -94,15 +94,24 @@ trait SanityCheckerParams extends DerivedFeatureFilterParams {
   final val maxCorrelation = new DoubleParam(
     parent = this, name = "maxCorrelation",
     doc = "Maximum correlation (absolute value) allowed between a feature in the feature vector and the label",
-    isValid = ParamValidators.inRange(lowerBound = 0.0, upperBound = 1.0, lowerInclusive = true, upperInclusive = true)
+    isValid = ParamValidators.inRange(lowerBound = -0.1, upperBound = 1.1, lowerInclusive = true, upperInclusive = true)
   )
   def setMaxCorrelation(value: Double): this.type = set(maxCorrelation, value)
   def getMaxCorrelation: Double = $(maxCorrelation)
 
+  final val maxFeatureCorr = new DoubleParam(
```
Comment (gerashegalov) [inline core/src/main/scala/com/salesforce/op/stages/impl/preparators/SanityChecker.scala:102]: better spell out Correlation
